dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java

```diff
@@ -0,0 +1,119 @@
+import org.elasticsearch.common.settings.ImmutableSettings.Builder;
+import org.elasticsearch.common.transport.InetSocketTransportAddress;
+import org.icgc.dcc.dcc.common.es.core.DocumentWriter;
+import org.icgc.dcc.dcc.common.es.impl.BulkProcessorListener;
+import org.icgc.dcc.dcc.common.es.impl.ClusterStateVerifier;
+import org.icgc.dcc.dcc.common.es.impl.EsDocumentWriter;
+import org.icgc.dcc.dcc.common.es.impl.IndexingState;
+
+@Slf4j
+@NoArgsConstructor(access = PRIVATE)
+public final class DocumentWriterFactory {
+
+	private static final Random RANDOM_GENERATOR = new Random();
+	private static final int BULK_ACTIONS = -1; // Unlimited
+
+	public static DocumentWriter createDocumentWriter(DocumentWriterConfiguration configuration) {
+		val client = newTransportClient(configuration.esUrl(), true);
+		val writerId = createWriterId();
+		val indexingState = new IndexingState(writerId);
+		val clusterStateVerifier = new ClusterStateVerifier(client, configuration.indexName(), writerId, indexingState);
+		val bulkProcessorListener = new BulkProcessorListener(clusterStateVerifier, indexingState, writerId);
+		val bulkProcessor = createProcessor(client, bulkProcessorListener);
+		return new EsDocumentWriter(client, configuration.indexName(), indexingState, bulkProcessor, writerId);
+	}
+
+	@SuppressWarnings("resource")
+	private static TransportClient newTransportClient(@NonNull String esUri, boolean sniff) {
+		val host = getHost(esUri);
+		val port = getPort(esUri);
+		val address = new InetSocketTransportAddress(host, port);
+
+		log.info("Creating ES transport client from URI '{}': host = '{}', port = {}", new Object[] { esUri, host, port });
+		return new TransportClient(createSettings(sniff)).addTransportAddress(address);
+	}
+
+	@SneakyThrows
+	private static String getHost(String esUri) {
+		return new URI(esUri).getHost();
+	}
+
+	@SneakyThrows
+	private static int getPort(String esUri) {
+		return new URI(esUri).getPort();
+	}
+
+	/**
+	 * Creates the client settings.
+	 * <p>
+	 * 
+	 * @see http://www.elasticsearch.org/guide/en/elasticsearch/client/java-api/current/client.htmls
+	 */
+	private static Builder createSettings(boolean sniff) {
+		return ImmutableSettings.settingsBuilder()
+
+		// Increase the ping timeout from the 5s (default) to something larger to prevent transient
+		// NoNodeAvailableExceptions
+				.put("client.transport.ping_timeout", "20s")
+
+				// The time to wait for a ping response from a node. Defaults to 5s.
+				.put("client.transport.nodes_sampler_interval", "10s")
+
+				// Enable / disable the client to sniff the rest of the cluster, and add those into its list of machines to use.
+				// In this case, note that the IP addresses used will be the ones that the other nodes were started with (the
+				// "publish" address)
+				.put("client.transport.sniff", sniff);
+	}
+
+	private static String createWriterId() {
+		val id = RANDOM_GENERATOR.nextInt(Integer.MAX_VALUE);
```

Comment:
Formatter seems to be using tabs. Can we reformat with the standard spaces? Thanks.